File: app/src/main/java/org/caojun/salmagundi/svgmap/PathItem.java
package org.caojun.salmagundi.svgmap;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.RectF;
import android.graphics.Region;

/**
 * @author: 蜗牛
 * @date: 2017/6/1 10:13
 * @desc:
 */

public class PathItem {
    private String title;
    private Path mPath;
    private boolean isSelect;

    public PathItem(String title, Path path) {
        this.title = title;
        mPath = path;
    }

    public String getMapName() {
        String name = title;
        String AND = " and ";
        String SPACE = " ";
        if (name.contains(AND)) {
            name = title.substring(0, title.indexOf(AND)) + "_" + title.substring(title.indexOf(AND) + AND.length());
        } else if (name.contains(SPACE)) {
            name = title.replace(SPACE, "_");
        }
        return name.toLowerCase();
    }

    public String getTitle() {
        return title;
    }

    /**
     * 是否touch在该path内部
     * @param x
     * @param y
     * @return
     */
    public boolean isTouch(int x, int y) {
        Region result = new Region();
        //构造一个区域对象。
        RectF r=new RectF();
        //计算path的边界
        mPath.computeBounds(r, true);
        //设置区域路径和剪辑描述的区域
        result.setPath(mPath, new Region((int)r.left,(int)r.top,(int)r.right,(int)r.bottom));
        return result.contains(x, y);
    }

    public void draw(Canvas canvas, Paint paint) {
        if (isSelect) {
            //首先绘制选中的背景阴影
            paint.clearShadowLayer();
            paint.setStyle(Paint.Style.FILL);
            paint.setShadowLayer(8, 0, 0, Color.BLACK);
            canvas.drawPath(mPath,paint);
            //绘制具体显示的
            paint.clearShadowLayer();
            paint.setColor(Color.BLUE);
            paint.setStyle(Paint.Style.FILL);
            canvas.drawPath(mPath,paint);
        } else {
            //绘制具体显示的
            paint.clearShadowLayer();
            paint.setColor(Color.GRAY);
            paint.setStyle(Paint.Style.FILL);
            canvas.drawPath(mPath,paint);
        }
    }

    public void setSelect(boolean select) {
        isSelect = select;
    }

    public RectF getRectF() {
        RectF r = new RectF();
        //计算path的边界
        mPath.computeBounds(r, true);
        return r;
    }
}

File: SvgMapLibrary/src/main/java/org/caojun/svgmap/PathItem.kt
package org.caojun.svgmap

import android.content.Context
import android.graphics.*

class PathItem {

    val context: Context
    val index: Int
    val id: String
    val title: String
    private var path: Path
    var isSelected = false
    private var colorSelected: Int = Color.BLUE
    private var colorUnselected: Int = Color.GRAY

    constructor(context: Context, index: Int, id: String, title: String, path: Path): this(context, index, id, title, path, Color.BLUE, Color.GRAY)

    constructor(context: Context, index: Int, id: String, title: String, path: Path, colorSelected: Int, colorUnselected: Int) {
        this.context = context
        this.index = index
        this.id = id.replace('-', '_').toLowerCase()
        this.title = title
        this.path = path
        this.colorSelected = colorSelected
        this.colorUnselected = colorUnselected
    }

    //地图名称（国名、地区名）
    fun getName(): String {
        val resId = if (id == "do") "${this.id}_" else id
        val id = getStringResId(context, resId)
        return if (id == null) {
            title
        } else {
            context.getString(id)
        }
    }

    //国旗（地区旗）图片Url
    fun getFlag(): Int {
        val flag = "flag_$id"
        return getDrawableResId(context, flag)?:0
    }

    /**
     * 是否touch在该path内部
     * @param x
     * @param y
     * @return
     */
    fun isTouch(x: Int, y: Int): Boolean {
        val result = Region()
        //构造一个区域对象。
        val r = RectF()
        //计算path的边界
        path.computeBounds(r, true)
        //设置区域路径和剪辑描述的区域
        result.setPath(path, Region(r.left.toInt(), r.top.toInt(), r.right.toInt(), r.bottom.toInt()))
        return result.contains(x, y)
    }

    fun draw(canvas: Canvas, paint: Paint) {
        paint.clearShadowLayer()
        if (isSelected) {
            //首先绘制选中的背景阴影
            paint.style = Paint.Style.FILL
            paint.setShadowLayer(8f, 0f, 0f, Color.BLACK)
            canvas.drawPath(path, paint)
            //绘制具体显示的
            paint.clearShadowLayer()
            paint.color = colorSelected
            paint.style = Paint.Style.FILL
        } else {
            //绘制具体显示的
            paint.color = colorUnselected
            paint.style = Paint.Style.FILL
        }
        canvas.drawPath(path, paint)
    }

    fun getRectF(): RectF {
        val r = RectF()
        //计算path的边界
        path.computeBounds(r, true)
        return r
    }

    private fun getResId(context: Context, name: String, defType: String): Int {
        return context.resources.getIdentifier(name, defType, context.packageName)
    }

    fun getRowResId(context: Context, name: String): Int? {
        val resId = getResId(context, name, "raw")
        return if (resId > 0) {
            resId
        } else {
            null
        }
    }

    fun getDrawableResId(context: Context, name: String): Int? {
        val resId = getResId(context, name, "drawable")
        return if (resId > 0) {
            resId
        } else {
            null
        }
    }

    fun getStringResId(context: Context, name: String): Int? {
        val resId = getResId(context, name, "string")
        return if (resId > 0) {
            resId
        } else {
            null
        }
    }
}